stdlib/Distributed/src/managers.jl

```diff
@@ -225,6 +268,82 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
 end
 
 
+function print_shell_escaped_windows(io, args::AbstractString...)::Nothing
+    first = true
+    quoted = false
+    for arg in args
+        first || write(io, ' ')
+        first = false
+        function isword(c::AbstractChar)
+            return 'a' <= c <= 'z' || 'A' <= c <= 'Z' || '0' <= c <= '9' ||
+                c == '\\' || c == '.' || c == '_' || c == '-' || c == ':' ||
+                c == '/' || c == '=' || c == '"' || c == '%'
+        end
+        quotes = !all(isword, arg) || arg == ""
+        if quotes; write(io, '"'); quoted = !quoted; end
+        backslashes = 0
+        for c in arg
+            if c == '\\'
+                backslashes += 1
+            else
+                if c == '"'
+                    backslashes = backslashes * 2 + 1
+                    quoted = !quoted
+                end
+                for j=1:backslashes
+                    write(io, '\\')
+                end
+                backslashes = 0
+                isword(c) || quoted || write(io, '^')   # for cmd.exe
```

Comment:
@mgkuhn  Could you explain this logic here with `isword`?  
For the `isword` function it looks like  every cmd.exe special meta-character is excluded except '"' and '%'. 

What's confusing is that the `^` escaping depends upon whether the argument is `quoted`, why? 

In the below example `$` is quoted and is not `isword`
 `shell_escape_win(`string\"\$`) |> print` gives `"string\"^$"`

but in this example this line is not triggered
 `shell_escape_win(`string\$`) |> print` gives `"string$"`

What's the rationale here? This is the only major thing I don't understand in the code. Everything else correctly implements the inverse of `CommandLineToArgvW `

Reply:
Welcome to [the wondrous world of the CMD.EXE parser](https://stackoverflow.com/questions/4094699/how-does-the-windows-command-interpreter-cmd-exe-parse-scripts). Whether `^` is an escape character or not depends on whether there have been an even or odd number of `"` on the line so far (see the Phase 2 description and the description of the quote flag in the previous link). We have not much control over the placement of `"`, as CMD.EXE always passes these on to the application, which then (this is typically done by libc according to the [conventions of the Microsoft C runtime](https://docs.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments)) uses them to split the command-line string into an argv array. Therefore we have to alternate between the two available quoting mechanisms. This all gets complicated by the fact that the `"` are seen and interpreted by both CMD.EXE and the application.

Reply:
@mgkuhn  Thanks for the reply. (BTW I opened a PR to get portions of this PR merged, only because I thought you  weren't interested in pushing this PR through due to inactivity. I'd be glad if you have the time otherwise, I can try to understand the issue and push it through.)

I think I kind of understand the gist of your comment. However wouldn't it be much easier to implement the following logic described at the end of the post here.  https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/

> 1 Always escape all arguments so that they will be decoded properly by CommandLineToArgvW, perhaps using my ArgvQuote function above.
> 
> 2 After step 1, then if and only if the command line produced will be interpreted by cmd, prefix each shell metacharacter (or each character) with a ^ character.
> 

Currently your code tries to intelligently infer if we should quote the argument depending if all the characters in the word are `isword`. Since we know each argument  is the output of `shell_split`, it doesn't matter  and we can just `\"arg\"`. This is much simpler. 

As consequence of step 1, we will always have an even number of `"`  and thus `^` will always be an escape character.

Reply:
In any case we can disregard my comment above, since it does indeed seem simpler to use your idea to toggle the state depending on `"` .  It seems that all the other processing should be properly taken care of by CMD, other than handling `^`.

Reply:
I'm still planning to push some variant of this PR through by the end of this year, but I will need more time for thorough testing as well as also covering PowerShell. I'm nore sure this PR is actually closely related to the one of Shell mode in REPL. The escaping routines in this PR are designed to pass strings transparently (past CMD.EXE or PowerShell) to an application (such as Julia) that parses the command line using the Microsoft C conventions. In REPL Shell mode, this is not the only or even main requirement. You may want to send commands that are interpreted by CMD.EXE, which parses its lines using completely different conventions from CommandLineToArgvW. I don't see how any function inverting CommandLineToArgvW would be of use in REPL Shell mode for Windows, as CMD.EXE does no use anything remotely resembling CommandLineToArgvW. I haven't looked yet closely into how REPL Shell mode works, but for Windows it would have to pass the whole command line transparently as a single string to the shell.